projects/js-packages/connection/components/in-place-connection/index.jsx

```diff
@@ -0,0 +1,128 @@
+/**
+ * External dependencies
+ */
+import React, { useEffect, useRef } from 'react';
+import { __ } from '@wordpress/i18n';
+import PropTypes from 'prop-types';
+
+/**
+ * Internal dependencies
+ */
+import './style.scss';
+
+/**
+ * The in-place connection component.
+ *
+ * @param {object} props -- The properties.
+ * @param {string} props.title -- Element title.
+ * @param {boolean} props.isLoading -- Whether the element is still loading.
+ * @param {string|number} props.width -- Iframe width.
+ * @param {string|number} props.height -- Iframe height.
+ * @param {boolean} props.hasConnectedOwner -- Whether the site has connection owner connected.
+ * @param {boolean} props.scrollToIframe -- Whether we need to auto-scroll the window upon element rendering.
+ * @param {string} props.connectUrl -- The connection URL.
+ * @param {Function} props.onComplete -- The callback to be called upon complete of the connection process.
+ * @param {string} props.source -- Component location identifier passed to WP.com.
+ *
+ * @returns {React.Component} The in-place connection component.
+ */
+const InPlaceConnection = props => {
+	const {
+		title,
+		isLoading,
+		width,
+		hasConnectedOwner,
+		scrollToIframe,
+		connectUrl,
+		onComplete,
+		source,
+	} = props;
+	let { height } = props;
+
+	const iframeWrapRef = useRef();
+	const iframeRef = useRef();
+
+	const receiveData = e => {
+		if ( e.source !== iframeRef.current.contentWindow ) {
+			return;
+		}
+
+		switch ( e.data ) {
+			case 'close':
+				// Remove listener, our job here is done.
+				window.removeEventListener( 'message', receiveData );
+
+				if ( onComplete ) {
+					onComplete();
+				}
+				break;
+			case 'wpcom_nocookie':
+				// Third-party cookies blocked. Let's redirect.
+				window.location.replace( connectUrl );
+				break;
+		}
+	};
+
+	useEffect( () => {
+		// Scroll to the iframe container
+		if ( scrollToIframe ) {
+			window.scrollTo( 0, iframeWrapRef.current.offsetTop - 10 );
+		}
+
+		// Add an event listener to identify successful authorization via iframe.
+		window.addEventListener( 'message', receiveData );
+	} );
+
+	// The URL looks like https://jetpack.wordpress.com/jetpack.authorize_iframe/1/. We need to include the trailing
+	// slash below so that we don't end up with something like /jetpack.authorize_iframe_iframe/
+	let src = connectUrl.replace( 'authorize/', 'authorize_iframe/' );
```

Comment:
For **future** reference: I never really liked this.. I think the connectUrl (at least the base) should be a const.
We want a certain iframe, we expect certain events from it as well (eg `wpcom_nocookie`) so ideally we should be enforcing it.

Furthermore, the logic is currently a bit messed. Ideally this component should rely on information we can fetch from the connection package (independent from Jetpack), however we currently get this from `build_authorize_url` (defined in `class.jetpack.php`).

We don't need to change it now, just something that troubles me..

Reply:
I agree, we need to fetch the required data from the connection package.
And I have started implementing this in another PR, so eventually the component will be more autonomous.
The `InPlaceConnection` component is a smaller part of it, and doesn't need to deal with fetching the data.